pkgs/unified_analytics/lib/src/asserts.dart

```diff
@@ -0,0 +1,126 @@
+/// Checks that the body of the request being sent to
+/// GA4 is within the limitations
+///
+/// Limitations can be found:
+/// https://developers.google.com/analytics/devguides/collection/protocol/ga4/sending-events?client_type=gtag#limitations
+
+void checkBody(Map<String, Object?> body) {
+  // Ensure we have the correct top level keys
+  if (!body.keys.contains('client_id')) {
+    throw AnalyticsException('client_id missing from top level keys');
+  }
+  if (!body.keys.contains('events')) {
+    throw AnalyticsException('events missing from top level keys');
+  }
+  if (!body.keys.contains('user_properties')) {
+    throw AnalyticsException('user_properties missing from top level keys');
+  }
+
+  final List events = body['events'] as List;
+  final Map<String, Object?> userProperties =
+      body['user_properties'] as Map<String, Object?>;
+
+  // GA4 Limitation:
+  // Requests can have a maximum of 25 events
+  if (events.length > 25) {
+    throw AnalyticsException('25 is the max number of events');
+  }
+
+  // Checks for each event object
+  for (Map<String, Object?> eventMap in events) {
+    // GA4 Limitation:
+    // Event names must be 40 characters or fewer, may only contain
+    // alpha-numeric characters and underscores, and must start
+    // with an alphabetic character
+    if ((eventMap['name'] as String).length > 40) {
+      throw AnalyticsException('Limit event names to 40 chars or less');
+    }
+    if (!RegExp(r'^[A-Za-z0-9_]+$').hasMatch(eventMap['name'] as String)) {
```

Comment:
Understood on part (b).

For part (a), would it be the same if I defined the pattern in outside of the function body in `asserts.dart`?

Reply:
Both the creation of the `RegExp` instance and the use of it to pattern match are expensive. By creating the instance statically so that it can be used repeatedly you minimize part of the expense. It doesn't matter (in terms of performance) whether it's a class field or a top-level variable.